p2p/net/swarm/black_hole_detector_test.go

```diff
@@ -0,0 +1,209 @@
+package swarm
+
+import (
+	"fmt"
+	"testing"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+func TestBlackHoleFilterReset(t *testing.T) {
+	n := 10
+	bhf := &blackHoleFilter{n: n, minSuccesses: 2, name: "test"}
+	var i = 0
+	// calls up to n should be probing
+	for i = 1; i <= n; i++ {
+		if bhf.HandleRequest() != blackHoleResultProbing {
+			t.Fatalf("expected calls up to n to be probes")
+		}
+		bhf.RecordResult(false)
+	}
+
+	// after threshold calls every nth call should be a probe
+	for i = n + 1; i < 42; i++ {
+		result := bhf.HandleRequest()
+		if (i%n == 0 && result != blackHoleResultProbing) || (i%n != 0 && result != blackHoleResultBlocked) {
+			t.Fatalf("expected every nth dial to be a probe")
+		}
+	}
+
+	bhf.RecordResult(true)
+	// check if calls up to n are probes again
+	for i = 0; i < n; i++ {
+		if bhf.HandleRequest() != blackHoleResultProbing {
+			t.Fatalf("expected black hole detector state to reset after success")
+		}
+		bhf.RecordResult(false)
+	}
+
+	// next call should be blocked
+	if bhf.HandleRequest() != blackHoleResultBlocked {
+		t.Fatalf("expected dial to be blocked")
+	}
+}
+
+func TestBlackHoleFilterSuccessFraction(t *testing.T) {
+	n := 10
+	tests := []struct {
+		minSuccesses, successes int
+		result                  blackHoleResult
+	}{
+		{minSuccesses: 5, successes: 5, result: blackHoleResultAllowed},
+		{minSuccesses: 3, successes: 3, result: blackHoleResultAllowed},
+		{minSuccesses: 5, successes: 4, result: blackHoleResultBlocked},
+		{minSuccesses: 5, successes: 7, result: blackHoleResultAllowed},
+		{minSuccesses: 3, successes: 1, result: blackHoleResultBlocked},
+		{minSuccesses: 0, successes: 0, result: blackHoleResultAllowed},
+		{minSuccesses: 10, successes: 10, result: blackHoleResultAllowed},
+	}
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("case-%d", i), func(t *testing.T) {
+			bhf := blackHoleFilter{n: n, minSuccesses: tc.minSuccesses}
+			for i := 0; i < tc.successes; i++ {
+				bhf.RecordResult(true)
+			}
+			for i := 0; i < n-tc.successes; i++ {
+				bhf.RecordResult(false)
+			}
+			got := bhf.HandleRequest()
+			if got != tc.result {
+				t.Fatalf("expected %d got %d", tc.result, got)
+			}
+		})
+	}
+}
+
+func areAddrsEqual(a, b []ma.Multiaddr) bool {
```

Comment:
Should we just use `require.ElementsMatch`?

Reply:
TIL this does deep equal. Thanks